src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+ * table.
+ */
+static void
+CreateShellTableForCitusLocalTable(List *shellTableDDLEvents)
+{
+	Assert(list_length(shellTableDDLEvents) > 0);
+
+	char *ddlCommand = NULL;
+	foreach_ptr(ddlCommand, shellTableDDLEvents)
+	{
+		StringInfo semicolonEndedCommand = makeStringInfo();
+		appendStringInfo(semicolonEndedCommand, "%s;", ddlCommand);
+
+		const char *commandString = semicolonEndedCommand->data;
+
+		Node *parseTree = ParseTreeNode(commandString);
+
+		/*
+		 * If the command defines a constraint, initially do not validate it
+		 * as shell table has no data in it. Note that this is only needed
+		 * before inserting metadata for citus local table. This is because,
+		 * after that, we already won't process constraints on the shell table.
+		 */
+		if (IsA(parseTree, AlterTableStmt))
+		{
+			AlterTableStmt *alterTableStmt = castNode(AlterTableStmt, parseTree);
+
+			/* we create ALTER TABLE commands so they only have one subcommand */
+			Assert(list_length(alterTableStmt->cmds) == 1);
+			AlterTableCmd *alterTableCmd =
+				castNode(AlterTableCmd, linitial(alterTableStmt->cmds));
+
+			if (alterTableCmd->subtype == AT_AddConstraint)
+			{
+				Constraint *constraint = (Constraint *) alterTableCmd->def;
+				constraint->skip_validation = true;
+			}
+		}
+
+		CitusProcessUtility(parseTree, commandString, PROCESS_UTILITY_TOPLEVEL,
+							NULL, None_Receiver, NULL);
+	}
+}
+
+
+/*
+ * InsertMetadataForCitusLocalTable inserts necessary metadata for the citus
+ * local table to the following metadata tables:
+ * pg_dist_partition, pg_dist_shard & pg_dist_placement.
+ */
+static void
+InsertMetadataForCitusLocalTable(Oid citusLocalTableId, uint64 shardId)
+{
+	Assert(OidIsValid(citusLocalTableId));
+	Assert(shardId != INVALID_SHARD_ID);
+
+	char distributionMethod = DISTRIBUTE_BY_NONE;
+	char replicationModel = ReplicationModel;
+
+	Assert(replicationModel != REPLICATION_MODEL_2PC);
+
+	uint32 colocationId = INVALID_COLOCATION_ID;
+	Var *distributionColumn = NULL;
+	InsertIntoPgDistPartition(citusLocalTableId, distributionMethod,
+							  distributionColumn, colocationId,
+							  replicationModel);
+
+	/* set shard storage type according to relation type */
+	char shardStorageType = ShardStorageType(citusLocalTableId);
+
+	text *shardMinValue = NULL;
+	text *shardMaxValue = NULL;
+	InsertShardRow(citusLocalTableId, shardId, shardStorageType,
+				   shardMinValue, shardMaxValue);
+
+	List *nodeList = list_make1(CoordinatorNode());
+
+	int replicationFactor = 1;
+	int workerStartIndex = 0;
+	InsertShardPlacementRows(citusLocalTableId, shardId, nodeList,
+							 workerStartIndex, replicationFactor);
+}
+
+
+/*
+ * IscitusLocalTable returns whether the given relationId identifies a citus
+ * local table.
+ */
+bool
+IsCitusLocalTable(Oid relationId)
+{
+	CitusTableCacheEntry *tableEntry = GetCitusTableCacheEntry(relationId);
+
+	if (!tableEntry->isCitusTable)
+	{
+		return false;
+	}
+
+	if (tableEntry->partitionMethod != DISTRIBUTE_BY_NONE)
```

Comment:
It also looks goods to me. But as I see, we less prefer long if checks in the codebase. (a similar function -> `IsReferenceTable`)

Reply:
I would still suggest using the above version because to me it seems harder to negotiate all the if checks to understand when the function returns true.

We can have the if check in multi lined version to make it more readable if that is a problem. What do you think?